tool/tsh/common/tlsmuxlistener.go

```diff
@@ -0,0 +1,93 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package common
+
+import (
+	"bufio"
+	"crypto/tls"
+	"net"
+
+	"github.com/gravitational/trace"
+)
+
+// connReader is a net.Conn wrapper with additional Peek() method.
+type connReader struct {
+	reader *bufio.Reader
+	net.Conn
+}
+
+// newBufferedConn is a connReader constructor.
+func newBufferedConn(conn net.Conn) connReader {
+	return connReader{
+		reader: bufio.NewReader(conn),
+		Conn:   conn,
+	}
+}
+
+// Peek reads n bytes without advancing the reader.
+// It's basically a wrapper around (bufio.Reader).Peek()
+func (b connReader) Peek(n int) ([]byte, error) {
+	return b.reader.Peek(n)
+}
+
+// Read returns data from underlying buffer.
+func (b connReader) Read(p []byte) (int, error) {
+	return b.reader.Read(p)
+}
+
+type tlsMuxListener struct {
```

Comment:
love the tls mux! could we move this to a util lib?

Reply:
It is somewhat experimental, but sure, we can move it.